File: backend/internals/models/ws.go
package models

type Action string

const (
	ActionJoinRoom    Action = "join_room"
	ActionLeaveRoom   Action = "leave_room"
	ActionSendMessage Action = "send_message"
)

type WSMessage struct {
	Action Action `json:"action"`
	Room   string `json:"room"`
	Data   string `json:"data"`
}

File: frontend/src/types/ws.ts
const WSAction = {
  JoinGame: "join_game",
} as const;

export type WSActionType = (typeof WSAction)[keyof typeof WSAction];

interface WSMessage<T = unknown> {
  action: WSActionType;
  room?: string;
  data?: T;
}

const emit = <T>(
  socket: WebSocket,
  action: WSActionType,
  room?: string,
  data?: T,
) => {
  const payload: WSMessage<T> = { action, room, data };
  socket.send(JSON.stringify(payload));
};

export { type WSMessage, WSAction, emit };
